// mongodb/src/cmap/auth/mongodb_oidc/token_entry_cache.ts
import type { IdPServerInfo, IdPServerResponse } from '../mongodb_oidc';
import { Cache } from './cache';

/* 5 minutes in milliseconds */
const EXPIRATION_BUFFER_MS = 300000;
/* Default expiration is now for when no expiration provided */
const DEFAULT_EXPIRATION_SECS = 0;
/** @internal */
export class TokenEntry {
  tokenResult: IdPServerResponse;
  serverInfo: IdPServerInfo;
  expiration: number;

  /**
   * Instantiate the entry.
   */
  constructor(tokenResult: IdPServerResponse, serverInfo: IdPServerInfo, expiration: number) {
    this.tokenResult = tokenResult;
    this.serverInfo = serverInfo;
    this.expiration = expiration;
  }

  /**
   * The entry is still valid if the expiration is more than
   * 5 minutes from the expiration time.
   */
  isValid() {
    return this.expiration - Date.now() > EXPIRATION_BUFFER_MS;
  }
}

/**
 * Cache of OIDC token entries.
 * @internal
 */
export class TokenEntryCache extends Cache<TokenEntry> {
  /**
   * Set an entry in the token cache.
   */
  addEntry(
    address: string,
    username: string,
    callbackHash: string,
    tokenResult: IdPServerResponse,
    serverInfo: IdPServerInfo
  ): TokenEntry {
    const entry = new TokenEntry(
      tokenResult,
      serverInfo,
      expirationTime(tokenResult.expiresIn
